pkg/cmd/roachtest/tests/mvcc_gc.go

```diff
@@ -0,0 +1,443 @@
+	r.Add(registry.TestSpec{
+		Name:    "mvcc_gc",
+		Owner:   registry.OwnerKV,
+		Timeout: 15 * time.Minute,
+		Cluster: r.MakeClusterSpec(3),
+		Run:     runMVCCGC,
+	})
+}
+
+/*
+Test verifies that GC respects range tombstones and garbage collects range
+tombstones and that mvcc stats are correctly updated.
+
+Test sequence:
+- create kv test data schema
+- pre-split with 3 min expiration to allow merging to proceed in parallel after
+  initial data is loaded
+- load test data using KV workload
+- repeat 3 times (while ranges are being gradually merged back)
+  - delete all data in kv table using range tombstones
+    (deletion is done using cli commands that sends kv requests to cluster)
+  - sets low GC ttl to make all data eligible for deletion after 2 min
+  - write more data over the existing one
+  - verify that all garbage and range keys were removed by GC within 5 min
+*/
+
+func runMVCCGC(ctx context.Context, t test.Test, c cluster.Cluster) {
+	rng, _ := randutil.NewTestRand()
+
+	// How many times test repeats generate/cleanup cycles after initial one.
+	const cleanupRuns = 3
+
+	c.Put(ctx, t.Cockroach(), "./cockroach")
+	s := install.MakeClusterSettings()
+	s.Env = append(s.Env, "COCKROACH_SCAN_INTERVAL=30s")
+	c.Start(ctx, t.L(), option.DefaultStartOpts(), s)
+
+	conn := c.Conn(ctx, t.L(), 1)
+	defer conn.Close()
+
+	execSQLOrFail := func(statement string, args ...interface{}) {
+		if _, err := conn.ExecContext(ctx, statement, args...); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	setClusterSetting := func(name string, value interface{}) {
+		execSQLOrFail(fmt.Sprintf(`SET CLUSTER SETTING %s = $1`, name), value)
+	}
+
+	setClusterSetting("storage.mvcc.range_tombstones.enabled", true)
+	setClusterSetting("kv.protectedts.poll_interval", "5s")
+	setClusterSetting("kv.mvcc_gc.queue_interval", "0s")
+
+	m := c.NewMonitor(ctx)
+	m.Go(func(ctx context.Context) error {
+		c.Run(ctx, c.Node(1), "./cockroach", "workload", "init", "kv", "--cycle-length", "10000")
+
+		t.L().Printf("finished init, doing force split ranges")
+		splitKvIntoRanges(t, conn, 100, 180*time.Second, "kv", "kv")
+
+		c.Run(ctx, c.Node(1), "./cockroach", "workload", "run", "kv", "--cycle-length", "10000",
+			"--max-block-bytes", "2048", "--min-block-bytes", "2048", "--duration", "2m",
+			"--read-percent", "0")
+
+		t.L().Printf("finished prepping initial GC data")
+
+		m := queryTableMetaOrFatal(t, conn, "kv", "kv")
+
+		for i := 0; i < cleanupRuns; i++ {
+			t.L().Printf("performing clean-assert cycle %d", i)
+
+			if err := retry.WithMaxAttempts(ctx, retry.Options{}, 3, func() error {
+				// This is not atomic, but should be ok if we lay more range tombstones if
+				// they still cover full range.
+				return deleteAllTableDataWithOverlappingTombstones(ctx, t, c, conn, rng, m, 5)
+			}); err != nil {
+				t.Fatal(err)
+			}
+
+			t.L().Printf("deleted all data using tombstones")
+
+			if err := checkTableHasNoLiveData(t, conn, "kv", "kv"); err != nil {
```

Comment:
Why do we assert this? It seems like it artificially limits the test space to full range deletions. Wouldn't test coverage be better if we had a mix of live data and garbage?

Reply:
We could either insert random set of point keys, deletes and range tombstones and then assert that range tombstones go away. We can't check that range tombstones have any effect by looking on stats e.g. we always have unmeasurable amount of garbage from overlapping keys and potentially point deletes which can't be separated from garbage covered by range tombstones.
Or we could write non overlapping data and delete all data with random set of range tombstones - they are misaligned with underlying ranges but always cover end to end. Then we could check that there's no live data after operation and once we collected them all there's no more garbage.

Reply:
> We can't check that range tombstones have any effect by looking on stats e.g. we always have unmeasurable amount of garbage from overlapping keys and potentially point deletes which can't be separated from garbage covered by range tombstones.

I'm not sure I understand. Is the concern whether range tombstones work at all, or whether we're writing them across any data? We have unit tests asserting that they work, and if we write them across the span where the existing data is written then we can be reasonably sure that they will cover data at least occasionally.

I don't think we should over-index on range tombstones here, and write a general GC roachtest that will ensure that all common variants of GC work as expected. But we could easily do both too, by first asserting that randomized GC works, and then at the very end write a single range tombstone across all of the data and assert that all data gets deleted and then GCed.